src/wasm/wasm-debug.cpp

```diff
@@ -99,26 +103,318 @@ void dumpDWARF(const Module& wasm) {
+
+  // Construct the map from the binaryLocations loaded from the wasm.
+  AddrExprMap(const Module& wasm) {
+    for (auto& func : wasm.functions) {
+      for (auto pair : func->binaryLocations) {
+        assert(map.count(pair.second) == 0);
+        map[pair.second] = pair.first;
+      }
+    }
+  }
+
+  // Construct the map from new binaryLocations just written
+  AddrExprMap(const BinaryLocationsMap& newLocations) {
+    for (auto pair : newLocations) {
+      assert(map.count(pair.second) == 0);
+      map[pair.second] = pair.first;
+    }
+  }
+
+  Expression* get(uint32_t addr) {
+    auto iter = map.find(addr);
+    if (iter != map.end()) {
+      return iter->second;
+    }
+    return nullptr;
+  }
+
+  void dump() {
+    std::cout << "  (size: " << map.size() << ")\n";
+    for (auto pair : map) {
+      std::cout << "  " << pair.first << " => " << pair.second << '\n';
+    }
+  }
+};
+
+static void updateDebugLines(const Module& wasm,
+                             llvm::DWARFYAML::Data& data,
+                             const BinaryLocationsMap& newLocations) {
+  // TODO: for memory efficiency, we may want to do this in a streaming manner,
+  //       binary to binary, without YAML IR.
+
+  // TODO: apparently DWARF offsets may be into the middle of instructions...
+  //       we may need to track their spans too
+  // https://github.com/WebAssembly/debugging/issues/9#issuecomment-567720872
+  AddrExprMap oldAddrMap(wasm);
+  // std::cout << "old\n";
+  // oldAddrMap.dump();
+  AddrExprMap newAddrMap(newLocations);
+  // std::cout << "new\n";
+  // newAddrMap.dump();
+
+  for (auto& table : data.DebugLines) {
+    // Parse the original opcodes and emit new ones.
+    LineState state(table);
+    // All the addresses we need to write out.
+    std::vector<uint32_t> newAddrs;
+    std::unordered_map<uint32_t, LineState> newAddrInfo;
+    for (auto& opcode : table.Opcodes) {
+      // Update the state, and check if we have a new row to emit.
+      if (state.update(opcode, table)) {
+        // An expression may not exist for this line table item, if we optimized
+        // it away.
+        if (auto* expr = oldAddrMap.get(state.addr)) {
+          auto iter = newLocations.find(expr);
+          if (iter != newLocations.end()) {
+            uint32_t newAddr = iter->second;
+            newAddrs.push_back(newAddr);
+            auto& updatedState = newAddrInfo[newAddr] = state;
+            // The only difference is the address TODO other stuff?
+            updatedState.addr = newAddr;
+          }
+        }
+        if (opcode.Opcode == 0 &&
+            opcode.SubOpcode == llvm::dwarf::DW_LNE_end_sequence) {
+          state = LineState(table);
+        }
+      }
+    }
+    // Sort the new addresses (which may be substantially different from the
+    // original layout after optimization).
+    std::sort(newAddrs.begin(), newAddrs.end(), [](uint32_t a, uint32_t b) {
+      return a < b;
```

Comment:
why is this custom comparator necessary when it's just uint32s?

Reply:
Oh right, the default order is ascending...